internal/pkg/agent/transpiler/vars.go

```diff
@@ -276,3 +294,15 @@ func varPrefixMatched(val string, key string) bool {
 	s := strings.SplitN(val, ".", 2)
 	return s[0] == key
 }
+
+func maybeAddDefaultProvider(val string, defaultProvider string) string {
```

Comment:
From a general solution standpoint, I assumed that the `defaultProvider` would come into play when resolving variables rather than when adding values: I was expecting that the resolver would first try resolving an expression like `${somevar.somefield.somestuff}` "as-is" and only in case of failure it would try to solve again by adding the default provider and resolving `${default_provider.somevar.somefield.somestuff}`.

What is being done here is modifying variable names that do not contain `"."` by prepending `defaultProvider.` when adding/replacing a variable and it only works for expressions that do not already contain "." which is a strong limitation.

Are there specific concerns/constraints that would make the current change the preferred solution?

Reply:
It cannot be done at resolve time because we need it to be known at `Observe` time. Without it at `Observe` time the code that was added to run only the providers that are referenced will not work. The issue is also present in the case that the variable references a fetch provider, as we will not be able to know if that variable is truly resolvable until that time occurs to resolve it.

If an expression contains `.` the text before the first `.` is always the provider, always! There is no other way to do this, because of the issues I placed above.

Reply:
> If an expression contains `.` the text before the first `.` is always the provider, always! There is no other way to do this, because of the issues I placed above.

Is this documented somewhere? I didn't see it in the PR changes but I didn't look at the whole `ast` package so I may have missed it. The reason I asked the question is that I suspected that there was a constraint somewhere, it wasn't obvious which and probably other people reading this snippet may have the same doubt.
It could be worth adding a comment on the function itself explaining that it will prepend the defaultProvider only on flat, single-token variable expression like `${somevar}`

Reply:
I added a doc string to the function to explain this better so it is known.